West docs

mbolivar opened the pull request:
Add initial documentation for West.

This covers its design and scope, current and future plans, and API documentation for the runner library.

CC: @tejlmand

Commit:
doc: conf.py: remove unused import

Make a pep8 linter happier.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: conf.py: make sure west is importable from Python

We need this to be able to pull in its API documentation.

This will need modifications when multi-repo support is available, in
order to get the imports from the independent West repo, but for now
just use the copy in Zephyr.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: add initial west documentation

This covers its design and scope, and provides usage and some
implementation documentation on the existing flashing and debugging
commands.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: conf.py: add sphinx.ext.autodoc extension

This will allow us to pull in Python API documentation from source
files.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented:
@galak better late than never, I hope :\

ulfalizer commented on doc/west/flash-debug.rst:
```diff
@@ -0,0 +1,243 @@
+programs running on a board: ``flash``, ``debug``, and
+``debugserver``.
+
+These use information stored in the CMake cache to flash or attach a
```
More of a comment, but as a CMake noob, I'd appreciate a short description of what the CMake cache is, or a link to the CMake documentation.

ulfalizer commented on doc/west/index.rst:
```diff
@@ -0,0 +1,94 @@
+
+- Bootloader integration: image building and signing, flashing
+  distinct images to different locations (the application is flashed
+  to separate area than the bootloader), and building and flashing the
```
"a separate area"?

"the application and the bootloader are flashed to separate areas" maybe.

mbolivar replied:
IMO that would make the subsequent clause ("and building and flashing the bootloader itself") a bit redundant. But I agree that this is clumsily written. I'll try to streamline it.

mbolivar replied:
Replaced with:

```
- Bootloader integration: bootloader-aware image building, signing,
  and flashing, as well as building and flashing the bootloader itself.
```

mbolivar commented on doc/conf.py:
```diff
@@ -15,9 +15,16 @@
 
+if "ZEPHYR_BASE" not in os.environ:
+    sys.stderr.write("$ZEPHYR_BASE environment variable undefined.\n")
+    exit(1)
```
That's a bit unrelated to this patch series, no? Note that I'm just moving this code higher in the file. I Can add a cleanup patch on top if you'd like.

ulfalizer replied:
Or just sneak it into the same patch I guess.

Commit:
conf.py: clean up exit if ZEPHYR_BASE is unset

Suggested by Ulf Magnusson.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented:
@ulfalizer thanks for review, comments addressed. I added a cleanup for the sys.exit shortcut, that's pretty neat!

carlescufi reviewed (approved):
Thanks Martí. This is great work. Very comprehensive and even covers the west design principles and constraints. I have no objections to the text as it stands.

mbolivar commented:
The true shippable failure is addressed in upstream west by https://github.com/zephyrproject-rtos/west/pull/4.

The other failure is spurious; there's no way I'm causing a QEMU riscv32 timing failure with this series.

Commit:
scripts: west: cherry-pick upstream 321ab2e17

Bring in the changes from upstream 321ab2e17 ("runner: core: fix
docstrings for ReST integration"). This is a minimal change necessary
to avoid documentation build issues. The rest of West has gained other
features that are too late to pull in to the RC period, so just bring
in what's needed to get to v1.12.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>
